### ISSUE TITLE
Speeding up CNN train time

### ISSUE THREAD

**Issue by cglagovich** [bamsumit/slayerPytorch#55]
First I want to say thanks for this repo, it's a great tool I have been experimenting with.

Working through the tutorials, I tried to train the N-MNIST CNN architecture on the full N-MNIST dataset. After creating a quicker dataset implementation (performs all read2Dspikes in __init__ and only does spikeToTensor in __getitem__), I obtained some training time results. It is starting to look like training a simple CNN on a dataset like N-MNIST would take incredibly long. Here are some quick results I got:

![image](https://user-images.githubusercontent.com/20099419/112345008-08c6d500-8c9b-11eb-8d96-39d4c3547126.png)

where the network architecture is as follows (num filters scaled by scaling factor in x axis of graph):

![image](https://user-images.githubusercontent.com/20099419/112345117-2300b300-8c9b-11eb-993d-a99c44b34f9c.png)

Does anyone have insights how to speed up training time for CNNs in SLAYER? I don't see how I can train a larger network on larger event images in reasonable time with this type of train time scaling. I would be happy to provide more information, via PyTorch profiling or whatever is required.

Specs:
GPU: RTX2060 Super, 8GB memory
RAM: 32 GB 
Batch size: 8
Train set size: 60,000
Test set size: 10,000

**Comment by bamsumit** [comment]
* Training SNN is going to be slower than ANN due to the inherent temporal dimension. You will be looking at a couple of days for training a reasonable sized network (appx. 3 days for the NMNIST CNN if I remember correctly).
* In most cases, dataloader is not the limiting bottleneck when you deploy it in multiple threads. PyTorch generally does pre-fetch of multiple samples in advance and have them ready to go in memory.

**Implementation tips:** 
* Define your fc1 layer like this instead
  `self.fc1 = slayer.dense(8*8*4*scalar, 10)`
  This will need explicit flatten operation in `forward()` before applying fc1
  `spike = spike.reshape(spike.shape[0], -1, 1, 1, spike.shape[-1])`
  But you will most likely find it to run faster than in the version above.
* For srm model, you can switch the order of synaptic operation and psp operation, i.e. `slayer.spike(slayer.psp(conv(spike)))` is same as `slayer.spike(conv(slayer.psp()))`. So you can be smart about implementing it. Apply `slayer.psp` on the dimension that is going to be smaller. **This does not apply to SLAYER-Loihi though.**
* Also I find that trying to fill up the GPU memory does not always give the max speedup. So it's advised to vary your batch size and look at avg throughput per sample. 

Hope that helps.

**Comment by cglagovich** [comment]
Thanks for taking the time to write this thoughtful response. This will really help me going forward. By the way, do you have a resource to learn more about the slayer.psp operation or the theory behind it?

You were correct about the dataloader, it was not the bottleneck after all.

Also, do you have the weights for the NMNIST CNN in the SLAYER paper publicly available? I don't know if you have/are able to release them but it would be interesting as I am trying to compare performance of SNN frameworks.

Thanks again for the help!

**Comment by bamsumit** [comment]
For SRM model, psp is convolution in time axis. slayer.psp uses the normalized psp response for a neuron and applies convolution in time axis.

The model in the paper was trained with CUDA implementation of SLAYER. I don't have the PyTorch version.

I am closing this issue, but let's keep this thread going. It will be helpful for other users.

**Comment by cglagovich** [comment]
I am back with some training time results, and some confusing early findings. Using a CNN on the full NMNIST set, I varied the input sample time and sample length (Ts and tSample in the YAML, respectively). I collected the time to train the network for one epoch for a few configurations. 

![image](https://user-images.githubusercontent.com/20099419/113881511-7dfdd400-978a-11eb-86a7-c21f45be11bf.png)

I thought that the training time would only be dependent on the number of bins we have in the input tensor, 
`nTimeBins = int(sample_length / sampling_time)`

However, we see that the main predictor of training time in the few samples I acquired was the Ts. Specifically, when sample_length=50 and sample_time=1 (nTimeBins=50), the training time is significantly higher than when sample_length=350 and sample_time=5 (nTimeBins=70). Do these results make sense? Without fully understanding the internals, this is not what I would have predicted.

Thanks @bamsumit !
